### DISCUSSION TITLE
PrivateAllocatorCapabilityState identity

### DISCUSSION THREAD

**Discussion by PhilDay-CT** [CHERIoT-Platform/cheriot-rtos#259] [Q&A] [answered]
Not sure if this is a bug or a clever piece of C magic ?

(https://github.com/microsoft/cheriot-rtos/blob/db4d635c46d4cb19e3bdf739ac94458c419f59f1/sdk/core/allocator/main.cc#L32C2-L38C4)

defines identifier as a uint16_t, but the value assigned to it is a uint32_t

https://github.com/microsoft/cheriot-rtos/blob/db4d635c46d4cb19e3bdf739ac94458c419f59f1/sdk/core/allocator/main.cc#L385C1-L394C1

The public version of the struct has some padding at the end, so I don't think it would break anything, but it looks odd ?

**Comment by davidchisnall** [comment, marked as answer]
The value will be truncated on assignment. Line 388 checks that it is a small enough value to be stored in the owner field of the chunk header, which is narrower than a 16-bit value and so, implicitly, that truncating to 16 bits will be sufficient. We store it as a 32-bit value so that we can do the overflow checks easily: if it overflows a 16-bit value by one, it’s still a valid 32-bit value and so the comparison on line 388 is safe and in range.

We should probably add:

 - A static assert check on that the chunk header owner ID width is narrower than (or equal width to) the field in the quota structure.
 - A comment explaining all of this.

**Comment by PhilDay-CT** [comment]
Thanks for the explanation.
